core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
   val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, types)
 
   /**
    * :: DeveloperApi ::
    * Implemented by subclasses to compute a given partition.
    */
   override def compute(split: Partition, context: TaskContext): Iterator[A] = {
+    new H2ORDDIterator(frameKeyName, split.index)
+  }
 
-    val iterator = new H2OChunkIterator[A] {
-
-      val jc = implicitly[ClassTag[A]].runtimeClass
-      val cs = jc.getConstructors
-      val ccr = cs.collectFirst({
-                case c if c.getParameterTypes.length == colNames.length => c
-              })
-        .getOrElse({
-            throw new IllegalArgumentException(
-                  s"Constructor must take exactly ${colNames.length} args")
-      })
-
-      override val keyName = frameKeyName
-      override val partIndex = split.index
-      val expectedTypes: Option[Array[Byte]] = expectedTypesAll
-
-      /* Converter context */
-      override val converterCtx: ReadConverterContext =
-      ConverterUtils.getReadConverterContext(isExternalBackend,
-        keyName,
-        chksLocation,
-        expectedTypes,
-        partIndex)
+  private val columnTypeNames = ReflectionUtils.typeNames[A](colNamesInResult)
+
+  private val jc = implicitly[ClassManifest[A]].runtimeClass
+
+  private type Deserializer = ReadConverterContext => Int => Any
+
+  private val DefaultDeserializer: Deserializer = _ => _ => null
+
+  private val plainExtractors: Map[String, Deserializer] = Map(
+    "Boolean" -> ((source: ReadConverterContext) => (col: Int) => source.getBoolean(col)),
+    "Byte"    -> ((source: ReadConverterContext) => (col: Int) => source.getByte(col)),
+    "Double"  -> ((source: ReadConverterContext) => (col: Int) => source.getDouble(col)),
+    "Float"   -> ((source: ReadConverterContext) => (col: Int) => source.getFloat(col)),
+    "Integer" -> ((source: ReadConverterContext) => (col: Int) => source.getInt(col)),
+    "Long"    -> ((source: ReadConverterContext) => (col: Int) => source.getLong(col)),
+    "Short"   -> ((source: ReadConverterContext) => (col: Int) => source.getShort(col)),
+    "String"  -> ((source: ReadConverterContext) => (col: Int) => source.getString(col)))
+
+  private def returnOption[X](op: ReadConverterContext => Int => X) = (source: ReadConverterContext) => (col: Int) => opt(op(source)(col))
+
+  type TypeName = String
+
+  private val allExtractors: Map[TypeName, Deserializer] = plainExtractors ++
+    (plainExtractors map {case (key, op) => s"Option[$key]" -> returnOption(op)})
+
+  private val extractorsMap: Map[TypeName, Deserializer] = allExtractors withDefaultValue DefaultDeserializer
+
+  val extractors = extractorsMap compose columnTypeNames
+
+  private def opt[X](op: => Any): Option[X] = try {
+    Option(op.asInstanceOf[X])
+  } catch {
+    case ex: Exception => None
+  }
+  // maps data columns to product components
+  val columnMapping: Map[Int, Int] =
```

Comment:
```
  val columnMapping: Map[Int, Int] =
    if (columnTypeNames.size == 1) Map(0->0) else multicolumnMapping
```
It is not the same thing. If there's one column, nobody cares about column names (judging by the existing tests)